### PR TITLE
Add CI using GitHub Actions

### PR DESCRIPTION
**Pull request by evanrelf**
This enables CI for this repository using GitHub Actions (which is free
for open source projects).

### CODE REVIEW

**Review by ixmatus — DISMISSED** [review]
We definitely want to pin at some point but getting builds back is a higher priority and thank you for doing this @evanrelf!!

**Comment by evanrelf** [comment]
Hmm, looks like there's an issue with the Nix build. I'll investigate a little later.

File: .github/workflows/ci.yml
Comment (riz0id) [inline .github/workflows/ci.yml:15]: We previously used [`scripts/fourmolu`](https://github.com/awakesecurity/spectacle/blob/main/scripts/fourmolu) to guard against merging unformatted code, I'm not sure if this would work:
```suggestion
      - run: nix-build release.nix --attr spectacle
  fourmolu:
    runs-on: ubuntu-latest
    needs: build
    steps:
      - uses: cachix/install-nix-action@v12
        name: Check
      - run: |
        sh nix-build --no-out-link --attr fourmolu ./release.nix
```

Comment (evanrelf) [reply .github/workflows/ci.yml:15]: I think the YAML might look a little different, but I can definitely add the `fourmolu` check 👍

**Comment by evanrelf** [comment]
For some reason, `haskell.packages.ghc8101` has `Cabal` at version 3.2 ([as expected](https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/libraries/version-history)), but `cabal-install` at version 3.0. This prevents me from performing a normal incremental build with `cabal-install` in the Nix shell, due to this error:

```
[nix-shell:~/Projects/awakesecurity/spectacle]$ cabal build
Warning: Unknown/unsupported 'ghc' version detected (Cabal 3.0.0.0 supports
'ghc' version < 8.10):
/nix/store/w8c5bpcg2q66c01spw88cg0xxvy96f3q-ghc-8.10.1-with-packages/bin/ghc
is version 8.10.1
```

I tried updating `cabal-install` to 3.2, but then it complains it needs `base` < 4.14 (GHC 8.10's `base`). So I jailbroke `cabal-install` and then it failed on some `base` related change.

I believe this explains (or is at least related to) the `Setup.hs` problems seen in CI. Choosing a different version of GHC than the default provided by `haskellPackages` is not a "happy path" in Nixpkgs, in my experience. Nothing is cached, and there are issues like this because it isn't finished being worked on, or not enough people use it, or whatever.

---

I can try switching from Nixpkgs 20.03 to 20.09 and trying again with GHC 8.10.1 (or preferably GHC 8.10.4), but that will also require going down a non-happy path of changing the default compiler.

The solution I would recommend is using a revision of the `master` branch of Nixpkgs, because it's already on GHC 8.10.4 as the default compiler. That means it's better supported, and everything is cached. It would probably also require less manual pinning to get newer libraries.

Thoughts?

**Comment by ixmatus** [comment]
@evanrelf we do not need to use the overridden GHC anymore now that we're not building with BuildKite. We could not use a newer nixpkgs with a GHC 8.10.x because we did not have a new enough version of Nix on our builders to support it. That is no longer a problem, so we do not need to use the "custom" GHC anymore.

**Comment by evanrelf** [comment]
Awesome, then I'll move to `master` then 👍

**Comment by evanrelf** [comment]
So...

Good news:

- Pinning Nixpkgs at a revision of `master` works great! It provides GHC 8.10.4, and everything is cached by `cache.nixos.org`
- `spectacle` is able to start building with _no pinned Haskell libraries_ (the only thing I have in the `haskellPackages` overlay is `callCabal2nix` for `spectacle`).

Bad news:

- The open source release of `spectacle` is missing lots of code (???). These files are present in the private repo, and are missing in the public repo:

  ```
  src/Data/Var.hs
  src/Language/Spectacle/Spec.hs
  test/Checker.hs
  test/Checker/Diehard.hs
  test/Lang/Sum.hs
  test/Spec.hs
  test/Spec/CyclicDependencies.hs
  test/Spec/Exceptions.hs
  test/Spec/Sharing.hs
  ```

- There are parts of `spectacle.cabal` that are wrong (e.g. missing `natural-transformation` dependency, incorrect module names in `exposed-modules`, etc.).

Here's the errors I got when trying to compile:

<details>
<summary>Logs</summary>

```
Building library for spectacle-0.1.0..
[ 1 of 27] Compiling Data.Fingerprint.Array ( src/Data/Fingerprint/Array.hs, dist/build/Data/Fingerprint/Array.o, dist/build/Data/Fingerprint/Array.dyn_o )
[ 2 of 27] Compiling Data.Fingerprint.Binary ( src/Data/Fingerprint/Binary.hs, dist/build/Data/Fingerprint/Binary.o, dist/build/Data/Fingerprint/Binary.dyn_o )
[ 3 of 27] Compiling Data.Fingerprint ( src/Data/Fingerprint.hs, dist/build/Data/Fingerprint.o, dist/build/Data/Fingerprint.dyn_o )
[ 4 of 27] Compiling Data.Type.HList.Internal ( src/Data/Type/HList/Internal.hs, dist/build/Data/Type/HList/Internal.o, dist/build/Data/Type/HList/Internal.dyn_o )
[ 5 of 27] Compiling Data.Type.HList.Class ( src/Data/Type/HList/Class.hs, dist/build/Data/Type/HList/Class.o, dist/build/Data/Type/HList/Class.dyn_o )
[ 6 of 27] Compiling Data.Type.HList  ( src/Data/Type/HList.hs, dist/build/Data/Type/HList.o, dist/build/Data/Type/HList.dyn_o )
[ 7 of 27] Compiling Data.Type.Induction ( src/Data/Type/Induction.hs, dist/build/Data/Type/Induction.o, dist/build/Data/Type/Induction.dyn_o )
[ 8 of 27] Compiling Language.Spectacle.Checker ( src/Language/Spectacle/Checker.hs, dist/build/Language/Spectacle/Checker.o, dist/build/Language/Spectacle/Checker.dyn_o )

src/Language/Spectacle/Checker.hs:36:1: error:
    Could not find module ‘Control.Monad.Mapping’
    Perhaps you meant
      Control.Monad.Fail (from base-4.14.1.0)
      Control.Monad.Morph (needs flag -package-key mmorph-1.1.5)
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
   |
36 | import Control.Monad.Mapping (MappingT)
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Language/Spectacle/Checker.hs:39:1: error:
    Could not find module ‘Language.Spectacle.Spec’
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
   |
39 | import Language.Spectacle.Spec (runSpec)
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ 9 of 27] Compiling Language.Spectacle.Lang.Sum ( src/Language/Spectacle/Lang/Sum.hs, dist/build/Language/Spectacle/Lang/Sum.o, dist/build/Language/Spectacle/Lang/Sum.dyn_o )
[10 of 27] Compiling Language.Spectacle.Syntax.Ascript ( src/Language/Spectacle/Syntax/Ascript.hs, dist/build/Language/Spectacle/Syntax/Ascript.o, dist/build/Language/Spectacle/Syntax/Ascript.dyn_o )
[11 of 27] Compiling Language.Spectacle.Syntax.Name ( src/Language/Spectacle/Syntax/Name.hs, dist/build/Language/Spectacle/Syntax/Name.o, dist/build/Language/Spectacle/Syntax/Name.dyn_o )
[12 of 27] Compiling Language.Spectacle.Syntax ( src/Language/Spectacle/Syntax.hs, dist/build/Language/Spectacle/Syntax.o, dist/build/Language/Spectacle/Syntax.dyn_o )

src/Language/Spectacle/Syntax.hs:9:43: error:
    Module
    ‘Language.Spectacle.Syntax.Ascript’
    does not export
    ‘type (#)((:#:))’
  |
9 | import Language.Spectacle.Syntax.Ascript (type (#) ((:#:)))
  |                                           ^^^^^^^^^^^^^^^^
[13 of 27] Compiling Language.Spectacle.RTS.Callstack ( src/Language/Spectacle/RTS/Callstack.hs, dist/build/Language/Spectacle/RTS/Callstack.o, dist/build/Language/Spectacle/RTS/Callstack.dyn_o )
[14 of 27] Compiling Language.Spectacle.Type.Rec ( src/Language/Spectacle/Type/Rec.hs, dist/build/Language/Spectacle/Type/Rec.o, dist/build/Language/Spectacle/Type/Rec.dyn_o )
[15 of 27] Compiling Data.FTCQueue    ( src/Data/FTCQueue.hs, dist/build/Data/FTCQueue.o, dist/build/Data/FTCQueue.dyn_o )
[16 of 27] Compiling Language.Spectacle.Lang.Control ( src/Language/Spectacle/Lang/Control.hs, dist/build/Language/Spectacle/Lang/Control.o, dist/build/Language/Spectacle/Lang/Control.dyn_o )
[17 of 27] Compiling Language.Spectacle.Lang.Internal ( src/Language/Spectacle/Lang/Internal.hs, dist/build/Language/Spectacle/Lang/Internal.o, dist/build/Language/Spectacle/Lang/Internal.dyn_o )
[18 of 27] Compiling Language.Spectacle.Lang.Interpreters ( src/Language/Spectacle/Lang/Interpreters.hs, dist/build/Language/Spectacle/Lang/Interpreters.o, dist/build/Language/Spectacle/Lang/Interpreters.dyn_o )
[19 of 27] Compiling Language.Spectacle.Lang ( src/Language/Spectacle/Lang.hs, dist/build/Language/Spectacle/Lang.o, dist/build/Language/Spectacle/Lang.dyn_o )
[20 of 27] Compiling Language.Spectacle.Syntax.Subst ( src/Language/Spectacle/Syntax/Subst.hs, dist/build/Language/Spectacle/Syntax/Subst.o, dist/build/Language/Spectacle/Syntax/Subst.dyn_o )
[21 of 27] Compiling Language.Spectacle.RTS.Exception ( src/Language/Spectacle/RTS/Exception.hs, dist/build/Language/Spectacle/RTS/Exception.o, dist/build/Language/Spectacle/RTS/Exception.dyn_o )
[22 of 27] Compiling Language.Spectacle.RTS ( src/Language/Spectacle/RTS.hs, dist/build/Language/Spectacle/RTS.o, dist/build/Language/Spectacle/RTS.dyn_o )
[23 of 27] Compiling Language.Spectacle.Effect.NonDet ( src/Language/Spectacle/Effect/NonDet.hs, dist/build/Language/Spectacle/Effect/NonDet.o, dist/build/Language/Spectacle/Effect/NonDet.dyn_o )
[24 of 27] Compiling Language.Spectacle.Syntax.Quantifier ( src/Language/Spectacle/Syntax/Quantifier.hs, dist/build/Language/Spectacle/Syntax/Quantifier.o, dist/build/Language/Spectacle/Syntax/Quantifier.dyn_o )
[25 of 27] Compiling Language.Spectacle.RTS.Env ( src/Language/Spectacle/RTS/Env.hs, dist/build/Language/Spectacle/RTS/Env.o, dist/build/Language/Spectacle/RTS/Env.dyn_o )
[26 of 27] Compiling Language.Spectacle.Syntax.Closure ( src/Language/Spectacle/Syntax/Closure.hs, dist/build/Language/Spectacle/Syntax/Closure.o, dist/build/Language/Spectacle/Syntax/Closure.dyn_o )
[27 of 27] Compiling Language.Spectacle.Syntax.Relation ( src/Language/Spectacle/Syntax/Relation.hs, dist/build/Language/Spectacle/Syntax/Relation.o, dist/build/Language/Spectacle/Syntax/Relation.dyn_o )
builder for '/nix/store/hixcx7y5dqc89v933s1r7gdhng8pjplq-spectacle-0.1.0.drv' failed with exit code 1
```

</details>

---

My other PR can be closed, since pinning things won't be necessary after using the newer Nixpkgs. But this PR is blocked until one of you guys restores the missing code.

File: .github/workflows/ci.yml
Comment (evanrelf) [reply .github/workflows/ci.yml:20]: Added in 379521545891252c96e15d292899ccc4770f24ff

**Comment by riz0id** [comment]
@evanrelf I crammed in a fairly large merge before mirroring the private repo to this one and updating the cabal was just missed in the final commits. #5 has the correct cabal file, natural-transformation can be added and the missing tests were intentionally removed by #1.

Apologies, I wasn't expecting CI to be added so quickly.

**Comment by evanrelf** [comment]
@riz0id No worries! 🙂 There's no rush IMO, it was just a little puzzling. I'll wait until you get #5 merged.

**Comment by riz0id** [comment]
Sounds good, thanks for taking the time to knock this out btw!

**Comment by evanrelf** [comment]
This PR has grown way larger than I initially anticipated; apologies for the scope creep. I'd appreciate it if you re-reviewed my changes, and if everything looks good (and CI passes), I strongly recommend _squash_ merging this.

**Review by riz0id — APPROVED** [review]

**Review by ixmatus — APPROVED** [review]

**Pull request merged**
